gdal/frmts/jpegxr/jpegxrdataset.cpp

```diff
@@ -0,0 +1,1202 @@
+        }
+        CPLAssert(eQuality != QUALITY_UNSET);
+        return eQuality;
+    }
+
+    Overlap GetOverlap() const
+    {
+        if (eOverlap == OVERLAP_UNSET)
+        {
+            JPEGXREncoderConfig* pThis = const_cast<JPEGXREncoderConfig*>(this);
+            if (GetQuality() == QUALITY_LOSSLESS)
+            {
+                pThis->eOverlap = OVERLAP_NONE;
+            }
+            else
+            {
+                // Image width must be at least 2 MB wide for subsampled chroma and
+                // two levels of overlap
+                if (fQuality >= 0.5F /*|| pImageEncode->uWidth < 2 * MB_WIDTH_PIXEL*/)
+                    pThis->eOverlap = OVERLAP_ONE;
+                else
+                    pThis->eOverlap = OVERLAP_TWO;
+            }
+        }
+        CPLAssert(eOverlap != OVERLAP_UNSET);
+        return eOverlap;
+    }
+
+    OVERLAP GetOVERLAP() const
+    {
+        Overlap ol = GetOverlap();
+        if (ol > OVERLAP_UNSET && ol < OVERLAP_COUNT)
+        {
+            return OVERLAP(static_cast<int>(ol));
+        }
+
+        throw JPEGXRException(WMP_errInvalidParameter);
+    }
+
+    Subsampling GetSubsampling() const
+    {
+        if (eSubsampling == SUBSAMPLING_UNSET)
+            const_cast<JPEGXREncoderConfig*>(this)->eSubsampling = SUBSAMPLING_444;
+        CPLAssert(eSubsampling != SUBSAMPLING_UNSET);
+        return eSubsampling;
+    }
+
+    void SetQuality(int n)
+    {
+        CPLAssert(QUALITY_LOWEST <= n && n <= QUALITY_LOSSLESS);
+        if (QUALITY_LOWEST <= n && n <= QUALITY_LOSSLESS)
+        {
+            fQuality = n / 100.0f;
+
+            if (fQuality > QUALITY_LOWEST && fQuality < QUALITY_LOSSLESS)
+                eQuality = QUALITY_CUSTOM;
+        }
+    }
+
+    void SetOverlap(int n)
+    {
+        CPLAssert(OVERLAP_UNSET < n && n < OVERLAP_COUNT);
+        if (OVERLAP_UNSET < n && n < OVERLAP_COUNT)
+            eOverlap = static_cast<Overlap>(n);
+    }
+
+    void SetSubsampling(int n)
+    {
+        CPLAssert(SUBSAMPLING_420 <= n && n < SUBSAMPLING_COUNT);
+        if (SUBSAMPLING_420 <= n && n < SUBSAMPLING_COUNT)
+            eSubsampling = static_cast<Subsampling>(n);
+    }
+
+public:
+    float fQuality = 1.0;
+
+private:
+    Quality eQuality = QUALITY_UNSET;
+    Overlap eOverlap = OVERLAP_UNSET;
+    Subsampling eSubsampling = SUBSAMPLING_UNSET;
+};
+
+/************************************************************************/
+/* ==================================================================== */
+/*                         JPEGXREncoder                                */
+/* ==================================================================== */
+/************************************************************************/
+
+struct JPEGXREncoder final
+{
+    friend struct JPEGXREncoderConfig;
+
+    JPEGXREncoder()
+        : pFactory(nullptr)
+        , pCodecFactory(nullptr)
+        , pImageEncode(nullptr)
+        , pEncodeStream(nullptr)
```

Comment:
@rouault No need. According to jxrlib code, encoder owns the stream and `PKImageEncode_Release` closes and releases it. I will add appropriate comment.